### PR TITLE
Lumberjack Powercreep (Added Chainsaws)

### PR DESCRIPTION
**Pull request by BurgerLUA**
https://www.youtube.com/watch?v=-YiUif84ig0

# Chainsaws
Yes, this is photoshopped because I don't know how to spawn a chainsaw and I locked myself out of the place where I put my chainsaw because I made a new character but it looks something like this.

Added Chainsaws. They can be held with two hands, and powered on. Sprites ripped from /tg/station. Mechanics ripped from chainsword + two handed weapon code. Sounds ripped from various websites,

## How to Use
Can be held with one or two hands. To powering it on requires it to be welded with two hands, however it can be held with one once powered on. Powering on is like a chainsaw IRL where RNG determines how many motor pulls are required to turn the thing on. Chainsaws require welder fuel, and can only be refilled at welder tanks.

## Damage
Chainsaws deal 10 brute damage on hit. They're bladed and sharp, mechanically. Chainsaws deal twice as much damage when turned on, and twice more when they're welded with two hands for a total of 40 damage. That's the same amount as a ninja's energy blade, except incredibly loud. You can hear a chainsaw about 2-3 screens away.

## Airlocks
Chainsaws can "cut through" bolts and open doors when given enough time and fuel. It's my own headcanon where airlocks can sense cuts and intrusion in the opening mechanism, and it opens in response thing it's an emergency responder with a special tool. It takes about 20 seconds to open a door, 40 if it's bolted, 40 if it's out of power, and 60 if it's out of power and bolted.

## How to Aquire
There is currently no way to acquire the chainsaw as is. I was thinking of making it an orderable cargo item given its power. In other codebases, the chainsaw can be built using a saw and some metal, however I think the chainsaw is too powerful for a makeshift weapon.

Perhaps a "Lumberjack Kit" for x amount of credits would work. Comes with a red flannel shirt, jeans, jackboots, a wool-knit hat, chainsaw, and some beer.

## Notes for admins
There is a special variant of the chainsaw called Bloody Chainsaw which holds a massive amount of fuel and performs door opening actions faster (5 seconds instead of the normal 20). Might be good for events.

## Final Notes
Suggestions and feedback are needed on the damage values for the Chainsaw and the acquiring of the chainsaw.

### CODE REVIEW

**Comment by TheGreatJorge** [comment]
Maybe add it as rare equipment from cargo warehouse, or heisters? Or antag item. Lumberjack kit seems kind of silly, given the setting. Though, I dunno.

Also, does it need some kind of resource to run? Like fuel, or something. That might be interesting as well.

**Comment by Arrow768** [comment]
Hm, what are you going to do with the chainsaw if you have it ?
Except cutting people that is.

**Comment by chaoko99** [comment]
Maybe more wood from logs?

**Comment by BurgerLUA** [comment]
Fuel seems like a good idea. Stores 300 units of welder fuel, emits Carbon Dioxide fumes when running.
I'll see what I can do about making it give more logs when used on wood.

**Comment by TheGreatJorge** [comment]
I mean. I guess it'd make good tool for antags. I heard torturing with chaisaw has it's benefits. Then exceptional riddance of vines and all that. And I additional wood for botany. Although I don't think botany ever has a shortage of it.

**Comment by BurgerLUA** [comment]
Added a lot of stuff this update.

- Chain pulling based on RNG
- Chain pulling sounds
- Shit-poor idle sounds
- Fuel requirement

**Comment by BurgerLUA** [comment]
Removed WIP title and added finishing touches.

**Comment by BurgerLUA** [comment]
Added WIP again because I think the chainsaw should be able to open lockers. Also new video in description.

File: code/game/machinery/doors/airlock.dm [hunk 1]
```diff
@@ -876,6 +876,34 @@ About the new airlock wires panel:
...
+		da.glass = 1
+	da.state = 1
+	da.created_name = src.name
+	da.update_state()
+	if(operating == -1 || (stat & BROKEN))
+		new /obj/item/weapon/circuitboard/broken(src.loc)
+		operating = 0
+	else
+		if (!electronics) create_electronics()
+		electronics.loc = src.loc
```
Comment (skull132) [inline code/game/machinery/doors/airlock.dm]: `electronics.forceMove(src.loc)`

File: code/game/objects/items/weapons/material/twohanded.dm [hunk 2]
```diff
@@ -277,3 +277,189 @@
...
+		var/mob/living/carbon/human/H = user
+		var/obj/item/organ/eyes/E = H.get_eyes()
+		if(!E)
+			return
+		var/safety = H.eyecheck()
+		if(H.status_flags & GODMODE)
+			return
+		switch(safety)
+			if(FLASH_PROTECTION_MODERATE)
+				usr << "<span class='warning'>Your eyes sting a little.</span>"
```
Comment (skull132) [inline code/game/objects/items/weapons/material/twohanded.dm]: `usr` is not allowed here. Do `user`

File: code/game/objects/items/weapons/material/twohanded.dm [hunk 3]
```diff
@@ -277,3 +277,189 @@
...
+				E.damage += rand(1, 2)*0.1
+				if(E.damage > 12)
+					user.eye_blurry += rand(3,6)*0.1
+			if(FLASH_PROTECTION_NONE)
+				usr << "<span class='warning'>Your eyes burn.</span>"
```
Comment (skull132) [inline code/game/objects/items/weapons/material/twohanded.dm]: `usr` is not allowed here. Do `user`

File: code/game/objects/items/weapons/material/twohanded.dm [hunk 4]
```diff
@@ -277,3 +277,189 @@
...
+				E.damage += rand(2, 4)*0.1
+				if(E.damage > 10)
+					E.damage += rand(4,10)*0.1
+			if(FLASH_PROTECTION_REDUCED)
+				usr << "<span class='danger'>Your equipment intensify the welder's glow. Your eyes itch and burn severely.</span>"
```
Comment (skull132) [inline code/game/objects/items/weapons/material/twohanded.dm]: `usr` is not allowed here. Do `user`

File: code/game/objects/items/weapons/material/twohanded.dm [hunk 5]
```diff
@@ -277,3 +277,189 @@
...
+		playsound(loc, 'sound/effects/refill.ogg', 50, 1, -6)
+		return
+	else if(powered)
+		playsound(loc, "sound/weapons/chainsword.ogg", 25, 0, 30)
+		RemoveFuel(3)
+		user.setClickCooldown(DEFAULT_ATTACK_COOLDOWN) //No sound spam
+
+	. = ..()
+
+/obj/item/weapon/material/twohanded/chainsaw/proc/eyecheck(mob/user as mob) //Shamefully copied from the welder. Damage values multiplied by 0.1
```
Comment (skull132) [inline code/game/objects/items/weapons/material/twohanded.dm]: cast user as `/mob/living/carbon/human/user` and you can write the opening as:
```dm
	if (!istype(user))
		return TRUE

	// your proc here, starting from line 387
```

**Comment by BurgerLUA** [comment]
tfw all your bad code is actually copy-pasted from already existing code

File: code/game/machinery/doors/airlock.dm [hunk 6]
```diff
@@ -876,6 +876,34 @@ About the new airlock wires panel:
...
+
+/obj/machinery/door/airlock/proc/CreateAssembly()
+	var/obj/structure/door_assembly/da = new assembly_type(src.loc)
+	if (istype(da, /obj/structure/door_assembly/multi_tile))
+		da.set_dir(src.dir)
+
+	da.anchored = 1
+	if(mineral)
+		da.glass = mineral
+	//else if(glass)
```
Comment (Arrow768) [inline code/game/machinery/doors/airlock.dm]: Minor Code-Quality nitpick.
Serves no purpose and is commented out -> Remove please

File: html/changelogs/burgerbb-chainsaw.yml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+author: BurgerBB
+delete-after: True
+changes: 
+  - rscadd: "Added chainsaws. Chainsaws are two handed weapons that require welder fuel to operate. Two hands a required to turn the chainsaw on. Chainsaws can open airlocks and lockers."
```
Comment (Arrow768) [inline html/changelogs/burgerbb-chainsaw.yml]: > Two hands a required to turn the chainsaw on. 

That is redundant. Its mentioned in the second sentence that chainsaws are two-handed.
Also, let them fiure out the part about airlocks and lockers themselfs (We dont need to give everything away)

**Pull request merged**